In [582]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt


####    Dicccionario de los datos

|   #   |   Titulo          |   Significado                         |
|   --- |    --------       |   -----------                         |
|   1   |   Year            |   Anio de fabricacion del vehiculo    |
|   2   |   make            |   Marca del vehiculo                  |
|   3   |   model           |   Modelo del vehiculo                 |
|   4   |   trim            |   Designacion adicional para el modelo del vehiculo   |
|   5   |   Body            |   Tipo de cuerpo del vehiculo         |
|   6   |   Transmition     |   tipo de transmision del vehiculo    |
|   7   |   vin             |   # numero de verificacion del vehiculo    |
|   8   |   state           |   Estado donde esta inscrito el vehiculo  |
|   9   |   condition       |   Condicion como se encuentra el vehiculo, en escala |
|   10  |   odometer        |   distancia de recorrido del vehivulo     |
|   11  |   color           |   Color del vehiculo                      |
|   12  |   interior        |   Color del interior del vehiculo         |
|   13  |   seller          |   La entidad que lo vende                 |
|   14  |   mmr             |   Informe del mercado de Manhein, sobre el precio referencial del vehiculoo   |
|   15  |   sellingprice    |   Precio de venta del vehivulo            |
|   16  |   saledate        |   Facha y hora de la venta                |

## importacion del dataset

In [583]:
ruta    =   '2.car_prices.csv'
df  =   pd.read_csv(ruta)

## previsualizacion dataset con la funcion head

In [584]:
df.head(3)

year make     model        trim   body transmission                vin  \
0  2015  Kia   Sorento          LX    SUV    automatic  5xyktca69fg566472   
1  2015  Kia   Sorento          LX    SUV    automatic  5xyktca69fg561319   
2  2014  BMW  3 Series  328i SULEV  Sedan    automatic  wba3c1c51ek116351   

  state  condition  odometer  color interior  \
0    ca        5.0   16639.0  white    black   
1    ca        5.0    9393.0  white    beige   
2    ca       45.0    1331.0   gray    black   

                                   seller      mmr  sellingprice  \
0                 kia motors america  inc  20500.0       21500.0   
1                 kia motors america  inc  20800.0       21500.0   
2  financial services remarketing (lease)  31900.0       30000.0   

                                  saledate  
0  Tue Dec 16 2014 12:30:00 GMT-0800 (PST)  
1  Tue Dec 16 2014 12:30:00 GMT-0800 (PST)  
2  Thu Jan 15 2015 04:30:00 GMT-0800 (PST)

# Revision de valores duplicados

In [585]:
fila, columna   =   df.shape
print(f'# de filas {fila}\n# de columnas {columna}')

# de filas 558837
# de columnas 16


In [586]:
df.drop_duplicates(subset=['vin','saledate'],keep='last',inplace=True)

In [587]:
fila, columna   =   df.shape
print(f'# de filas {fila}\n# de columnas {columna}')

# de filas 558746
# de columnas 16


Se analiza la columna 'vin' y 'saledate' debido a que vin es la identificacion del vehiculo, el cual es unico para cada vehiculo del registro y saledate porque solo es posible que se haga una ejecucion de la venta en el mismo dia y hora

In [588]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 558746 entries, 0 to 558836
Data columns (total 16 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   year          558746 non-null  int64  
 1   make          548449 non-null  object 
 2   model         548351 non-null  object 
 3   trim          548099 non-null  object 
 4   body          545556 non-null  object 
 5   transmission  493414 non-null  object 
 6   vin           558742 non-null  object 
 7   state         558746 non-null  object 
 8   condition     546948 non-null  float64
 9   odometer      558659 non-null  float64
 10  color         558005 non-null  object 
 11  interior      558005 non-null  object 
 12  seller        558746 non-null  object 
 13  mmr           558712 non-null  float64
 14  sellingprice  558734 non-null  float64
 15  saledate      558734 non-null  object 
dtypes: float64(4), int64(1), object(11)
memory usage: 72.5+ MB


## Revision de valores nulos o NaN
 Se presenta una gran cantidad de valores NaN entre la segunda y la sexta columna, tambien en la octava

In [589]:
df.isna().sum()

year                0
make            10297
model           10395
trim            10647
body            13190
transmission    65332
vin                 4
state               0
condition       11798
odometer           87
color             741
interior          741
seller              0
mmr                34
sellingprice       12
saledate           12
dtype: int64

In [590]:
(df.isna().sum()/fila)*100

year             0.000000
make             1.842877
model            1.860416
trim             1.905517
body             2.360643
transmission    11.692612
vin              0.000716
state            0.000000
condition        2.111514
odometer         0.015571
color            0.132618
interior         0.132618
seller           0.000000
mmr              0.006085
sellingprice     0.002148
saledate         0.002148
dtype: float64

Al sacar el porcentaje que representan los valores NaN de la cantidad de la data, vemos que la columna que se encuentra más comprometida es **transmission** ya que tiene un total de 11.69% de valores NaN, el resto de las columnas tienen menos del 5%

## Inspeccion de valores en las columnas

### Columna 1

In [591]:
df['year'].value_counts().sort_index()

year
1982         2
1983         1
1984         5
1985        10
1986        11
1987         8
1988        11
1989        20
1990        49
1991        67
1992       132
1993       205
1994       392
1995       710
1996       851
1997      1544
1998      2148
1999      3361
2000      5221
2001      6465
2002      9712
2003     13275
2004     17334
2005     21384
2006     26908
2007     30838
2008     31497
2009     20592
2010     26480
2011     48545
2012    102303
2013     98164
2014     81068
2015      9433
Name: count, dtype: int64

Esta columna no presenta valores NaN y tambien se observa que sus valores son consecuentes y no presenta valores "extranios"

### Columna 2

In [592]:
df['make'].sort_values().unique()

array(['Acura', 'Aston Martin', 'Audi', 'BMW', 'Bentley', 'Buick',
       'Cadillac', 'Chevrolet', 'Chrysler', 'Daewoo', 'Dodge', 'FIAT',
       'Ferrari', 'Fisker', 'Ford', 'GMC', 'Geo', 'HUMMER', 'Honda',
       'Hyundai', 'Infiniti', 'Isuzu', 'Jaguar', 'Jeep', 'Kia',
       'Lamborghini', 'Land Rover', 'Lexus', 'Lincoln', 'Lotus', 'MINI',
       'Maserati', 'Mazda', 'Mercedes-Benz', 'Mercury', 'Mitsubishi',
       'Nissan', 'Oldsmobile', 'Plymouth', 'Pontiac', 'Porsche', 'Ram',
       'Rolls-Royce', 'Saab', 'Saturn', 'Scion', 'Subaru', 'Suzuki',
       'Tesla', 'Toyota', 'Volkswagen', 'Volvo', 'acura', 'airstream',
       'audi', 'bmw', 'buick', 'cadillac', 'chev truck', 'chevrolet',
       'chrysler', 'dodge', 'dodge tk', 'dot', 'ford', 'ford tk',
       'ford truck', 'gmc', 'gmc truck', 'honda', 'hyundai', 'hyundai tk',
       'jeep', 'kia', 'land rover', 'landrover', 'lexus', 'lincoln',
       'maserati', 'mazda', 'mazda tk', 'mercedes', 'mercedes-b',
       'mercury', 'mitsubish

In [593]:
df['make'].str.upper().sort_values().unique()


array(['ACURA', 'AIRSTREAM', 'ASTON MARTIN', 'AUDI', 'BENTLEY', 'BMW',
       'BUICK', 'CADILLAC', 'CHEV TRUCK', 'CHEVROLET', 'CHRYSLER',
       'DAEWOO', 'DODGE', 'DODGE TK', 'DOT', 'FERRARI', 'FIAT', 'FISKER',
       'FORD', 'FORD TK', 'FORD TRUCK', 'GEO', 'GMC', 'GMC TRUCK',
       'HONDA', 'HUMMER', 'HYUNDAI', 'HYUNDAI TK', 'INFINITI', 'ISUZU',
       'JAGUAR', 'JEEP', 'KIA', 'LAMBORGHINI', 'LAND ROVER', 'LANDROVER',
       'LEXUS', 'LINCOLN', 'LOTUS', 'MASERATI', 'MAZDA', 'MAZDA TK',
       'MERCEDES', 'MERCEDES-B', 'MERCEDES-BENZ', 'MERCURY', 'MINI',
       'MITSUBISHI', 'NISSAN', 'OLDSMOBILE', 'PLYMOUTH', 'PONTIAC',
       'PORSCHE', 'RAM', 'ROLLS-ROYCE', 'SAAB', 'SATURN', 'SCION',
       'SMART', 'SUBARU', 'SUZUKI', 'TESLA', 'TOYOTA', 'VOLKSWAGEN',
       'VOLVO', 'VW', nan], dtype=object)

- Se puede observar que aparece la marca de carro Dodge y Dodge TK que al final vendrian ser las mismas, lo mismo con DOT 
- Lo mismo se presenta con Ford, Ford TK y fork truck
- GMC y GMC truck
- Hyundai y hyndai TK
- MAZDA y MAZDA TK
- MERCEDES, MERCEDES-B, MERCEDES-BENZ 
- LAND ROVER Y LANDROVER

Para homogeneizar la busqueda de estos valores, la columnas 'make' la convertiremos en mayuscula

In [594]:
df['make']  =   df['make'].str.upper()

cambio_dic1 =   {
                 'DODGE TK':'DODGE'  ,   'DOT':'DODGE',
                 'FORD TK':'FORD'   ,   'FORD TRUCK':'FORD',
                 'GMC TRUCK':'GMC',
                 'HYUNDAI TK':'HYUNDAI',
                 'MAZDA TK':'MAZDA',
                 'MERCEDES-B':'MERCEDES'    ,   'MERCEDES-BENZ':'MERCEDES',
                 'LANDROVER':'LAND ROVER'
                 }


df.replace({'make':cambio_dic1}, inplace=True)

df['make'].sort_values().unique()

array(['ACURA', 'AIRSTREAM', 'ASTON MARTIN', 'AUDI', 'BENTLEY', 'BMW',
       'BUICK', 'CADILLAC', 'CHEV TRUCK', 'CHEVROLET', 'CHRYSLER',
       'DAEWOO', 'DODGE', 'FERRARI', 'FIAT', 'FISKER', 'FORD', 'GEO',
       'GMC', 'HONDA', 'HUMMER', 'HYUNDAI', 'INFINITI', 'ISUZU', 'JAGUAR',
       'JEEP', 'KIA', 'LAMBORGHINI', 'LAND ROVER', 'LEXUS', 'LINCOLN',
       'LOTUS', 'MASERATI', 'MAZDA', 'MERCEDES', 'MERCURY', 'MINI',
       'MITSUBISHI', 'NISSAN', 'OLDSMOBILE', 'PLYMOUTH', 'PONTIAC',
       'PORSCHE', 'RAM', 'ROLLS-ROYCE', 'SAAB', 'SATURN', 'SCION',
       'SMART', 'SUBARU', 'SUZUKI', 'TESLA', 'TOYOTA', 'VOLKSWAGEN',
       'VOLVO', 'VW', nan], dtype=object)

In [595]:
df['trim']  =   df['trim'].str.upper()
df['model']  =   df['model'].str.upper()

In [596]:
df['codigo_marca']=df['vin'].str[:2]
df['codigo_modelo']=df['vin'].str[2:4]
df['codigo_transmision']=df['vin'].str[4:5]
df['codigo_trim']=df['vin'].str[5:7]
df['codigo']=df['vin'].str[:8]

df


year    make                model         trim       body  \
0       2015     KIA              SORENTO           LX        SUV   
1       2015     KIA              SORENTO           LX        SUV   
2       2014     BMW             3 SERIES   328I SULEV      Sedan   
3       2015   VOLVO                  S60           T5      Sedan   
4       2014     BMW  6 SERIES GRAN COUPE         650I      Sedan   
...      ...     ...                  ...          ...        ...   
558832  2015     KIA                 K900       LUXURY      Sedan   
558833  2012     RAM                 2500  POWER WAGON   Crew Cab   
558834  2012     BMW                   X5    XDRIVE35D        SUV   
558835  2015  NISSAN               ALTIMA        2.5 S      sedan   
558836  2014    FORD                F-150          XLT  SuperCrew   

       transmission                vin state  condition  odometer  ...  \
0         automatic  5xyktca69fg566472    ca        5.0   16639.0  ...   
1         automatic  5xyktca69fg561319    ca        5.0    9393.0  ...   
2         automatic  wba3c1c51ek116351    ca       45.0    1331.0  ...   
3         automatic  yv1612tb4f1310987    ca       41.0   14282.0  ...   
4         automatic  wba6b2c57ed129731    ca       43.0    2641.0  ...   
...             ...                ...   ...        ...       ...  ...   
558832          NaN  knalw4d4xf6019304    in       45.0   18255.0  ...   
558833    automatic  3c6td5et6cg112407    wa        5.0   54393.0  ...   
558834    automatic  5uxzw0c58cl668465    ca       48.0   50561.0  ...   
558835    automatic  1n4al3ap0fc216050    ga       38.0   16658.0  ...   
558836    automatic  1ftfw1et2eke87277    ca       34.0   15008.0  ...   

       interior                                             seller      mmr  \
0         black                            kia motors america  inc  20500.0   
1         beige                            kia motors america  inc  20800.0   
2         black             financial services remarketing (lease)  31900.0   
3         black                            volvo na rep/world omni  27500.0   
4         black             financial services remarketing (lease)  66000.0   
...         ...                                                ...      ...   
558832    black                                   avis corporation  35300.0   
558833    black                                    i -5 uhlmann rv  30200.0   
558834    black             financial services remarketing (lease)  29800.0   
558835    black  enterprise vehicle exchange / tra / rental / t...  15100.0   
558836     gray                   ford motor credit company llc pd  29600.0   

        sellingprice                                 saledate codigo_marca  \
0            21500.0  Tue Dec 16 2014 12:30:00 GMT-0800 (PST)           5x   
1            21500.0  Tue Dec 16 2014 12:30:00 GMT-0800 (PST)           5x   
2            30000.0  Thu Jan 15 2015 04:30:00 GMT-0800 (PST)           wb   
3            27750.0  Thu Jan 29 2015 04:30:00 GMT-0800 (PST)           yv   
4            67000.0  Thu Dec 18 2014 12:30:00 GMT-0800 (PST)           wb   
...              ...                                      ...          ...   
558832       33000.0  Thu Jul 09 2015 07:00:00 GMT-0700 (PDT)           kn   
558833       30800.0  Wed Jul 08 2015 09:30:00 GMT-0700 (PDT)           3c   
558834       34000.0  Wed Jul 08 2015 09:30:00 GMT-0700 (PDT)           5u   
558835       11100.0  Thu Jul 09 2015 06:45:00 GMT-0700 (PDT)           1n   
558836       26700.0  Thu May 28 2015 05:30:00 GMT-0700 (PDT)           1f   

       codigo_modelo codigo_transmision codigo_trim    codigo  
0                 yk                  t          ca  5xyktca6  
1                 yk                  t          ca  5xyktca6  
2                 a3                  c          1c  wba3c1c5  
3                 16                  1          2t  yv1612tb  
4                 a6                  b          2c  wba6b2c5  
...              ...

In [597]:
df[(df['codigo']=='yv1612t')  ] #& (df['model']=='V8 Vantage')

Empty DataFrame
Columns: [year, make, model, trim, body, transmission, vin, state, condition, odometer, color, interior, seller, mmr, sellingprice, saledate, codigo_marca, codigo_modelo, codigo_transmision, codigo_trim, codigo]
Index: []

[0 rows x 21 columns]

In [598]:
df_1    =   df[:]
df_1.dropna(subset=['make','model','trim'], inplace=True)
resumen    =   df_1.groupby(['codigo','make','model','trim'])['seller'].count().reset_index()
resumen.sort_values(['codigo','seller'],ascending=[True, False]).drop_duplicates(subset=['codigo'],keep='first')

C:\Users\herna\AppData\Local\Temp\ipykernel_37172\472799380.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_1.dropna(subset=['make','model','trim'], inplace=True)


codigo         make       model    trim  seller
0      137za843       HUMMER          H1   WAGON       1
1      19uua564        ACURA          TL     3.2      35
2      19uua565        ACURA          TL     3.2       6
3      19uua566        ACURA          TL     3.2     187
4      19uua567        ACURA          TL     3.2      33
...         ...          ...         ...     ...     ...
14244  zff67nfa      FERRARI  458 ITALIA    BASE       1
14245  zffew58a      FERRARI        F430    BASE       1
14246  zffew59a      FERRARI        F430  SPIDER       2
14248  zffyt53a      FERRARI         360  SPIDER       1
14249  zhwgu22t  LAMBORGHINI    GALLARDO  SPYDER       4

[10574 rows x 5 columns]

In [599]:
resumen['codigo'].values

array(['137za843', '19uua564', '19uua565', ..., 'zffew59a', 'zffyt53a',
       'zhwgu22t'], dtype=object)

In [600]:
df_1    =df.head(0)
for x in lista_make:
    df_e    =   df[df['make']==x]
    df_r    =   df_e[['codigo','model','trim','seller']]
    df_r.dropna(subset=['model','trim'], inplace=True)
    df_r    =   df_r.groupby(['codigo','model','trim'])['seller'].count().reset_index()
    df_r    =   df_r.sort_values(['codigo','seller'],ascending=[True, False]).drop_duplicates(subset=['codigo'],keep='first')

    #for y in df_r['codigo'].values     :

        #(df_e[df_e['codigo']==y])['model']=(df_r[df_r['codigo']==y])['model']
        #(df_e[df_e['trim']==y])['model']=(df_r[df_r['trim']==y])['model']
    df_e_inner      =   pd.merge( df_e   ,    df_r   ,   on='codigo',how='inner', suffixes=('_old', ''))
    
    df_e_left       =   df_e[~df_e['codigo'].isin(list(df_e_inner['codigo'].values))]
    df_e            =   pd.concat([df_e_inner,df_e_left])
    
    df_1    =   pd.concat([df_1,df_e])


C:\Users\herna\AppData\Local\Temp\ipykernel_37172\2813750724.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_r.dropna(subset=['model','trim'], inplace=True)
C:\Users\herna\AppData\Local\Temp\ipykernel_37172\2813750724.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_r.dropna(subset=['model','trim'], inplace=True)
C:\Users\herna\AppData\Local\Temp\ipykernel_37172\2813750724.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


C:\Users\herna\AppData\Local\Temp\ipykernel_37172\2813750724.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_r.dropna(subset=['model','trim'], inplace=True)
C:\Users\herna\AppData\Local\Temp\ipykernel_37172\2813750724.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_r.dropna(subset=['model','trim'], inplace=True)
C:\Users\herna\AppData\Local\Temp\ipykernel_37172\2813750724.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


In [601]:
df_1

year   make   model                trim   body transmission  \
0   2014  ACURA     ILX  TECHNOLOGY PACKAGE  Sedan    automatic   
1   2013  ACURA     ILX  TECHNOLOGY PACKAGE  Sedan          NaN   
2   2013  ACURA     ILX  TECHNOLOGY PACKAGE  Sedan          NaN   
3   2013  ACURA     ILX  TECHNOLOGY PACKAGE  Sedan          NaN   
4   2013  ACURA     ILX  TECHNOLOGY PACKAGE  Sedan    automatic   
..   ...    ...     ...                 ...    ...          ...   
19  2014     VW  ROUTAN             SEL PRM    NaN    automatic   
20  2014     VW  ROUTAN             SEL PRM    NaN    automatic   
21  2014     VW  BEETLE             4C 1.8T    NaN    automatic   
22  2014     VW  ROUTAN                   S    NaN    automatic   
23  2000     VW   JETTA             GLS TDI    NaN       manual   

                  vin state  condition  odometer  ... sellingprice  \
0   19vde1f70ee008913    ca       25.0    9051.0  ...      21250.0   
1   19vde1f70de001314    pa       37.0   30920.0  ...      19300.0   
2   19vde1f79de012716    pa       42.0   23552.0  ...      22000.0   
3   19vde1f79de006091    pa        2.0   14057.0  ...      18800.0   
4   19vde1f72de001752    nj       39.0    7101.0  ...      21500.0   
..                ...   ...        ...       ...  ...          ...   
19  2c4rvadg5er134795    oh       43.0   25053.0  ...      25000.0   
20  2c4rvadg6er135373    il       46.0   18769.0  ...      24600.0   
21  3vwf17atxem658290    fl       47.0    6357.0  ...      11800.0   
22  2c4rvaagxer144680    il        5.0   14998.0  ...      17600.0   
23  3vwsf29m9ym121947    wa        2.0  156603.0  ...       2500.0   

                                   saledate codigo_marca  codigo_modelo  \
0   Thu Dec 18 2014 12:00:00 GMT-0800 (PST)           19             vd   
1   Fri Dec 19 2014 09:00:00 GMT-0800 (PST)           19             vd   
2   Thu Dec 18 2014 09:30:00 GMT-0800 (PST)           19             vd   
3   Fri Jan 02 2015 09:00:00 GMT-0800 (PST)           19             vd   
4   Wed Jan 07 2015 09:30:00 GMT-0800 (PST)           19             vd   
..                                      ...          ...            ...   
19  Tue Jan 06 2015 01:30:00 GMT-0800 (PST)           2c             4r   
20  Tue Jan 13 2015 10:00:00 GMT-0800 (PST)           2c             4r   
21  Tue Jan 06 2015 09:30:00 GMT-0800 (PST)           3v             wf   
22  Tue Jan 13 2015 10:00:00 GMT-0800 (PST)           2c             4r   
23  Wed Jan 14 2015 05:30:00 GMT-0800 (PST)           3v             ws   

    codigo_transmision codigo_trim    codigo model_old            trim_old  \
0                    e          1f  19vde1f7       ILX  TECHNOLOGY PACKAGE   
1                    e          1f  19vde1f7       ILX  TECHNOLOGY PACKAGE   
2                    e          1f  19vde1f7       ILX  TECHNOLOGY PACKAGE   
3                    e          1f  19vde1f7       ILX  TECHNOLOGY PACKAGE   
4                    e          1f  19vde1f7       ILX  TECHNOLOGY PACKAGE   
..                 ...         ...       ...       ...                 ...   
19                   v          ad  2c4rvadg    ROUTAN             SEL PRM   
20                   v          ad  2c4rvadg    ROUTAN             SEL PRM   
21                   1          7a  3vwf17at    BEETLE             4C 1.8T   
22                   v          aa  2c4rvaag    ROUTAN                   S   
23                   f          29  3vwsf29m     JETTA             GLS TDI   

                                     seller_old  
0                  american honda motor company  
1   ahfc/honda lease/trust/hvt  inc. eot acuras  
2                          tsj auto brokers inc  
3                       piazza acura of reading  
4                      super buy auto sales inc  
..                                          ...  
19                                    vw credit  
20                       vw credit prequalified  
21                 florida auto financial group  
22                  

In [652]:
df.isna().sum()

year                      0
make                  10297
model                 10395
trim                  10647
body                  13190
transmission          65332
vin                       4
state                     0
condition             11798
odometer                 87
color                   741
interior                741
seller                    0
mmr                      34
sellingprice             12
saledate                 12
codigo_marca              4
codigo_modelo             4
codigo_transmision        4
codigo_trim               4
codigo                    4
dtype: int64

In [653]:
df.shape

(558746, 21)

In [656]:
df[df['make'].isna()]

year make model trim body transmission                vin state  \
742     2012  NaN   NaN  NaN  NaN    automatic  wbakb8c51cc964387    ca   
747     2012  NaN   NaN  NaN  NaN    automatic  wbakb8c53cc964410    ca   
766     2012  NaN   NaN  NaN  NaN    automatic  wbakb8c54cc964089    ca   
798     2012  NaN   NaN  NaN  NaN    automatic  wbakb8c59cc448049    ca   
803     2012  NaN   NaN  NaN  NaN    automatic  wbakb8c58cc962863    ca   
...      ...  ...   ...  ...  ...          ...                ...   ...   
558545  2000  NaN   NaN  NaN  NaN    automatic  4n2xn11txyd844711    ma   
558617  2007  NaN   NaN  NaN  NaN    automatic  jtjbt20x770129305    pa   
558736  2011  NaN   NaN  NaN  NaN    automatic  4a4jn2as6be029938    nv   
558773  2005  NaN   NaN  NaN  NaN    automatic  1g1yy24u355116011    tx   
558776  2013  NaN   NaN  NaN  NaN    automatic  zam39nka2d0068288    ga   

        condition  odometer  ... interior  \
742          38.0   23208.0  ...    black   
747          33.0   19785.0  ...     gray   
766          37.0   48424.0  ...    black   
798          48.0   39825.0  ...     gray   
803          49.0   35093.0  ...      tan   
...           ...       ...  ...      ...   
558545       31.0  108900.0  ...     gray   
558617       41.0   66393.0  ...    beige   
558736       41.0   67820.0  ...    black   
558773       24.0  114787.0  ...    black   
558776       42.0   30148.0  ...    black   

                                        seller      mmr  sellingprice  \
742     financial services remarketing (lease)  47200.0       46000.0   
747     financial services remarketing (lease)  49500.0       46000.0   
766     financial services remarketing (lease)  42300.0       43000.0   
798     financial services remarketing (lease)  58100.0       58500.0   
803     financial services remarketing (lease)  45200.0       44500.0   
...                                        ...      ...           ...   
558545                 boch toyota/scion south   1275.0        1100.0   
558617            r hollenshead auto sales inc  18150.0       18500.0   
558736                          imperial rides   9175.0       10500.0   
558773                    hopper motorplex inc  15400.0       14800.0   
558776              maserati north america inc  58500.0       47000.0   

                                       saledate codigo_marca codigo_modelo  \
742     Thu Feb 26 2015 04:30:00 GMT-0800 (PST)           wb            ak   
747     Thu Feb 12 2015 04:30:00 GMT-0800 (PST)           wb            ak   
766     Thu Jan 15 2015 04:30:00 GMT-0800 (PST)           wb            ak   
798     Thu Jan 15 2015 04:30:00 GMT-0800 (PST)           wb            ak   
803     Thu Jan 29 2015 04:30:00 GMT-0800 (PST)           wb            ak   
...                                         ...          ...           ...   
558545  Thu Jun 18 2015 11:30:00 GMT-0700 (PDT)           4n            2x   
558617  Fri Jun 19 2015 02:00:00 GMT-0700 (PDT)           jt            jb   
558736  Fri Jun 19 2015 05:00:00 GMT-0700 (PDT)           4a            4j   
558773  Wed Jun 24 2015 03:15:00 GMT-0700 (PDT)           1g            1y   
558776  Wed Jun 24 2015 03:00:00 GMT-0700 (PDT)           za            m3   

       codigo_transmision codigo_trim    codigo  
742                     b          8c  wbakb8c5  
747                     b          8c  wbakb8c5  
766                     b          8c  wbakb8c5  
798                     b          8c  wbakb8c5  
803                     b          8c  wbakb8c5  
...                   ...         ...       ...  
558545                  n          11  4n2xn11t  
558617                  t          20  jtjbt20x  
558736                  n          2a  4a4jn2as  
558773                  y          24  1g1yy24u  
558776                  9          nk  zam39nka  

[10297 rows x 21 columns]

In [661]:
rr=   df.dropna(subset=['make']).drop_duplicates(subset=['make', 'codigo'])
rr.groupby('codigo')['make'].count().sort_values()

codigo
137za843    1
5nmskdag    1
5npdh4ae    1
5npe24aa    1
5npe24af    1
           ..
3d73y4cl    2
3d73y3hl    2
3d73y3cl    2
1d7rv1cp    2
1d7rb1cp    2
Name: make, Length: 10607, dtype: int64

In [665]:
rr[rr['codigo']=='1g1yy24u']

year       make     model trim body transmission                vin  \
7236  2005  CHEVROLET  CORVETTE  NaN  NaN    automatic  1g1yy24ux55104289   

     state  condition  odometer  ... interior               seller      mmr  \
7236    ca       47.0   47908.0  ...    black  aaero sweet company  20900.0   

      sellingprice                                 saledate codigo_marca  \
7236       22750.0  Thu Dec 18 2014 11:30:00 GMT-0800 (PST)           1g   

     codigo_modelo codigo_transmision codigo_trim    codigo  
7236            1y                  y          24  1g1yy24u  

[1 rows x 21 columns]

In [654]:
df_1.shape

(548449, 24)

In [651]:
df_1.isna().sum()

year                      0
make                      0
model                    12
trim                    118
body                   2893
transmission          63572
vin                       4
state                     0
condition             11712
odometer                 84
color                   726
interior                726
seller                    0
mmr                      34
sellingprice             12
saledate                 12
codigo_marca              4
codigo_modelo             4
codigo_transmision        4
codigo_trim               4
codigo                    4
model_old               220
trim_old                366
seller_old              134
dtype: int64

In [602]:
import pandas as pd

# Suponiendo que tienes dos DataFrames llamados df1 y df2
# Donde df1 tiene 5 columnas y df2 tiene 3 columnas, con una columna 'ID' en ambos DataFrames

# Ejemplo de DataFrame df1
df1 = pd.DataFrame({
    'ID': [1, 2, 3, 4],
    'A': [10, 20, 30, 40],
    'B': [11, 21, 31, 41],
    'C': [12, 22, 32, 42],
    'D': [13, 23, 33, 43],
    'E': [14, 24, 34, 44]
})

# Ejemplo de DataFrame df2
df2 = pd.DataFrame({
    'ID': [2, 3, 4, 5],
    'A': [100, 200, 300, 400],
    'B': [110, 210, 310, 410],
    'C': [120, 220, 320, 420]
})

# Realizar el reemplazo basado en la columna 'ID'
df_result = pd.merge(df1, df2[['ID', 'A', 'B', 'C']], on='ID',  how='outer', suffixes=('_old', ''))

# Puedes seleccionar solo las columnas que deseas mantener
df_result = df_result[['ID', 'A', 'B', 'C', 'D', 'E']]

# Visualizar el resultado
print(df_result)


   ID      A      B      C     D     E
0   1    NaN    NaN    NaN  13.0  14.0
1   2  100.0  110.0  120.0  23.0  24.0
2   3  200.0  210.0  220.0  33.0  34.0
3   4  300.0  310.0  320.0  43.0  44.0
4   5  400.0  410.0  420.0   NaN   NaN


In [603]:
df_1[df_1['codigo']=='yv1612t']

Empty DataFrame
Columns: [year, make, model, trim, body, transmission, vin, state, condition, odometer, color, interior, seller, mmr, sellingprice, saledate, codigo_marca, codigo_modelo, codigo_transmision, codigo_trim, codigo, model_old, trim_old, seller_old]
Index: []

[0 rows x 24 columns]

In [604]:
element =   df[['make','codigo_modelo']].drop_duplicates(keep='first').sort_values('make')
element =   element[~element['make'].isna()]
element =   element[~element['codigo_modelo'].isna()]
element

make codigo_modelo
4840   ACURA            4k
745    ACURA            4c
808    ACURA            8t
856    ACURA            ny
23602  ACURA            he
...      ...           ...
65502     VW            wd
29094     VW            wc
26583     VW            ws
96078     VW            wf
69154     VW            4r

[952 rows x 2 columns]

In [605]:
element['codigo_modelo'].unique().shape

(555,)

In [606]:
element[element.duplicated(subset=['codigo_modelo'],keep=False)].sort_values('codigo_modelo')

make codigo_modelo
235818      FORD            0a
665      PORSCHE            0a
5928        AUDI            19
608        VOLVO            19
405     INFINITI            1a
...          ...           ...
389         MINI            wz
1142       HONDA            xf
45           BMW            xf
166920    NISSAN            zb
406298     MAZDA            zb

[621 rows x 2 columns]

Se presenta la correction de los datos de la columna 2

### Columna 3

Para presentar una homogeneidad en el formato, se colocara en mayuscula la columna 'model'

In [607]:
df['model'] =   df['model'].str.upper()


In [608]:
df['model'].unique().shape

(852,)

Existe 852 valores unicos 

-----------------------

In [609]:
df['model'] =   df['model'].str.upper()


Revisando la columna model, podemos observar que algunos modelos solo presentan un registro, esto se puede presentar por dos razones:
1. Es un modelo escaso (antiguo)
2. Su nombre esta escrito de manera erronea(Error humano)

In [610]:
modelo_cantida_valores  =   df['model'].value_counts(ascending=True)
modelo_cantida_valores

model
458 ITALIA         1
COMM               1
CTS-V WAGON        1
328I               1
42C                1
               ...  
ESCAPE         11893
CAMRY          12547
FUSION         12946
F-150          14475
ALTIMA         19347
Name: count, Length: 851, dtype: int64

In [667]:
col3_make   =   df_1.groupby(['make','model'])[['seller']].agg('count')
col3_make

seller
make  model          
ACURA CL           97
      EL            2
      ILX         140
      INTEGRA      44
      LEGEND        9
...               ...
VOLVO XC70        169
      XC90        734
VW    BEETLE        3
      JETTA         9
      ROUTAN       12

[818 rows x 1 columns]

ahora, a partir de este modelo, vamos a extraer los modelos que tienen menor coincidencia

In [668]:
col3_make   =   col3_make.reset_index()
col3_make.rename(columns={'seller':'cantidad'})

make    model  cantidad
0    ACURA       CL        97
1    ACURA       EL         2
2    ACURA      ILX       140
3    ACURA  INTEGRA        44
4    ACURA   LEGEND         9
..     ...      ...       ...
813  VOLVO     XC70       169
814  VOLVO     XC90       734
815     VW   BEETLE         3
816     VW    JETTA         9
817     VW   ROUTAN        12

[818 rows x 3 columns]

In [669]:
lista_make  =   list(col3_make['make'].unique())
len(lista_make)

56

#### Correccion de modelos de vehiculos

##### 1. Acura

In [670]:
col3_make[col3_make['make']==lista_make[0]]

make            model  seller
0   ACURA               CL      97
1   ACURA               EL       2
2   ACURA              ILX     140
3   ACURA          INTEGRA      44
4   ACURA           LEGEND       9
5   ACURA              MDX    1584
6   ACURA              RDX     420
7   ACURA               RL     131
8   ACURA              RLX      17
9   ACURA              RSX     151
10  ACURA               TL    2135
11  ACURA              TLX       1
12  ACURA              TSX    1119
13  ACURA  TSX SPORT WAGON      36
14  ACURA              ZDX      39

#### 2. AIRSTREAM

In [671]:
col3_make[col3_make['make']==lista_make[1]]

make       model  seller
15  AIRSTREAM  INTERSTATE       1

#### 3. ASTON MARTIN	

In [672]:
col3_make[col3_make['make']==lista_make[2]]

make       model  seller
16  ASTON MARTIN         DB9       6
17  ASTON MARTIN      RAPIDE       2
18  ASTON MARTIN  V8 VANTAGE      17

#### 4. AUDI

In [673]:
col3_make[col3_make['make']==lista_make[3]]

make            model  seller
19  AUDI               A3     259
20  AUDI               A4    2272
21  AUDI               A5     351
22  AUDI               A6     833
23  AUDI               A7     133
24  AUDI               A8     298
25  AUDI          ALLROAD      46
26  AUDI  ALLROAD QUATTRO      35
27  AUDI        CABRIOLET       1
28  AUDI               Q3       1
29  AUDI               Q5     542
30  AUDI               Q7     480
31  AUDI               R8      38
32  AUDI             RS 4       3
33  AUDI             RS 5      15
34  AUDI             RS 6       8
35  AUDI             RS 7       8
36  AUDI               S4     197
37  AUDI               S5     146
38  AUDI               S6      25
39  AUDI               S7       8
40  AUDI               S8      19
41  AUDI              SQ5       7
42  AUDI               TT     130
43  AUDI            TT RS       3
44  AUDI              TTS       6

#### 5. BENTLEY

In [674]:
col3_make[col3_make['make']==lista_make[4]]

make                          model  seller
45  BENTLEY        CONTINENTAL FLYING SPUR      31
46  BENTLEY  CONTINENTAL FLYING SPUR SPEED       2
47  BENTLEY                 CONTINENTAL GT      43
48  BENTLEY           CONTINENTAL GT SPEED       3
49  BENTLEY                CONTINENTAL GTC      33
50  BENTLEY          CONTINENTAL GTC SPEED       1
51  BENTLEY        CONTINENTAL SUPERSPORTS       1
52  BENTLEY                    FLYING SPUR       2

#### 6. BMW

In [675]:
col3_make[col3_make['make']==lista_make[5]]

make                  model  seller
53  BMW                      1       1
54  BMW               1 SERIES     404
55  BMW               2 SERIES      32
56  BMW               3 SERIES    8204
57  BMW  3 SERIES GRAN TURISMO      73
58  BMW                   320I       1
59  BMW                   323I       1
60  BMW                   328I       1
61  BMW               4 SERIES     293
62  BMW    4 SERIES GRAN COUPE      12
63  BMW               5 SERIES    4174
64  BMW  5 SERIES GRAN TURISMO     204
65  BMW               6 SERIES     676
66  BMW    6 SERIES GRAN COUPE     114
67  BMW               7 SERIES    1286
68  BMW                   750I      13
69  BMW                  750LI      38
70  BMW               8 SERIES       1
71  BMW         ACTIVEHYBRID 7      13
72  BMW        ACTIVEHYBRID X6       5
73  BMW                     I8       9
74  BMW                      M       6
75  BMW                     M3     297
76  BMW                     M4      10
77  BMW                     M5     107
78  BMW                     M6      83
79  BMW          M6 GRAN COUPE      15
80  BMW                     X1     331
81  BMW                     X3    1251
82  BMW                     X4       5
83  BMW                     X5    2462
84  BMW                   X5 M      37
85  BMW                     X6     263
86  BMW                   X6 M      26
87  BMW                     Z3      61
88  BMW                     Z4     276
89  BMW                   Z4 M       6

In [676]:
col3_make[col3_make['make']==lista_make[6]]

make        model  seller
90   BUICK       ALLURE       1
91   BUICK      CENTURY     227
92   BUICK      ENCLAVE    1070
93   BUICK       ENCORE      93
94   BUICK     LACROSSE    1189
95   BUICK      LESABRE     496
96   BUICK      LUCERNE     496
97   BUICK  PARK AVENUE     116
98   BUICK      RAINIER      95
99   BUICK        REGAL     603
100  BUICK   RENDEZVOUS     287
101  BUICK      RIVIERA       4
102  BUICK   ROADMASTER       3
103  BUICK      TERRAZA      55
104  BUICK       VERANO     382

In [677]:
col3_make[col3_make['make']==lista_make[7]]

make            model  seller
105  CADILLAC              ATS     201
106  CADILLAC           CATERA      29
107  CADILLAC              CTS    1934
108  CADILLAC        CTS COUPE     158
109  CADILLAC        CTS WAGON      14
110  CADILLAC            CTS-V      41
111  CADILLAC      CTS-V COUPE      35
112  CADILLAC      CTS-V WAGON       1
113  CADILLAC          DEVILLE     624
114  CADILLAC              DTS     522
115  CADILLAC         ELDORADO      55
116  CADILLAC              ELR       5
117  CADILLAC         ESCALADE    1184
118  CADILLAC     ESCALADE ESV     625
119  CADILLAC     ESCALADE EXT     154
120  CADILLAC  ESCALADE HYBRID      40
121  CADILLAC        FLEETWOOD       1
122  CADILLAC          SEVILLE     157
123  CADILLAC              SRX    1328
124  CADILLAC              STS     332
125  CADILLAC            STS-V       6
126  CADILLAC              XLR      26
127  CADILLAC              XTS     155

In [678]:
col3_make[col3_make['make']==lista_make[8]]

make model  seller
128  CHEV TRUCK   S10       1

In [679]:
col3_make[col3_make['make']==lista_make[9]]['model'].values

array(['1500', '2500', '3500', 'ASTRO', 'ASTRO CARGO', 'AVALANCHE',
       'AVEO', 'BLACK DIAMOND AVALANCHE', 'BLAZER', 'C/K 1500 SERIES',
       'C/K 2500 SERIES', 'C/K 3500 SERIES', 'CAMARO', 'CAPRICE',
       'CAPTIVA SPORT', 'CAVALIER', 'CLASSIC', 'COBALT', 'COLORADO',
       'CORSICA', 'CORVETTE', 'CORVETTE STINGRAY', 'CRUZE', 'EQUINOX',
       'EXPRESS', 'EXPRESS CARGO', 'G1500', 'G3500', 'HHR', 'IMPALA',
       'IMPALA LIMITED', 'LUMINA', 'MALIBU', 'MALIBU CLASSIC',
       'MALIBU HYBRID', 'MALIBU MAXX', 'MONTE CARLO', 'OPTRA', 'PRIZM',
       'S-10', 'S-10 BLAZER', 'S10', 'SILVERADO 1500',
       'SILVERADO 1500 CLASSIC', 'SILVERADO 1500 HYBRID',
       'SILVERADO 1500HD', 'SILVERADO 2500', 'SILVERADO 2500HD',
       'SILVERADO 2500HD CLASSIC', 'SILVERADO 3500', 'SILVERADO 3500HD',
       'SONIC', 'SPARK', 'SPARK EV', 'SS', 'SSR', 'SUBURBAN', 'TAHOE',
       'TAHOE HYBRID', 'TRACKER', 'TRAILBLAZER', 'TRAILBLAZER EXT',
       'TRAVERSE', 'UPLANDER', 'UPLANDR', 'VENTURE', 'VOLT']

In [680]:
col3_make[col3_make['make']==lista_make[10]]

make             model  seller
196  CHRYSLER               200    4049
197  CHRYSLER               300    3906
198  CHRYSLER              300M     101
199  CHRYSLER             ASPEN     228
200  CHRYSLER            CIRRUS      26
201  CHRYSLER          CONCORDE      69
202  CHRYSLER         CROSSFIRE      61
203  CHRYSLER             GRAND       1
204  CHRYSLER          LE BARON       1
205  CHRYSLER               LHS      21
206  CHRYSLER          PACIFICA     377
207  CHRYSLER           PROWLER       2
208  CHRYSLER                PT       1
209  CHRYSLER        PT CRUISER    1231
210  CHRYSLER           SEBRING    1358
211  CHRYSLER              TOWN      27
212  CHRYSLER  TOWN AND COUNTRY    5999
213  CHRYSLER         TWN/CNTRY       3
214  CHRYSLER           VOYAGER      23

In [681]:
col3_make[col3_make['make']==lista_make[11]]

make    model  seller
215  DAEWOO  LEGANZA       1
216  DAEWOO   NUBIRA       2

In [707]:
col3_make[col3_make['make']==lista_make[12]]

make            model  seller
217  DODGE          AVENGER    3963
218  DODGE            B1500       6
219  DODGE          CALIBER    1334
220  DODGE          CARAVAN     253
221  DODGE       CHALLENGER    1196
222  DODGE          CHARGER    4542
223  DODGE           DAKOTA     609
224  DODGE             DART    1028
225  DODGE          DURANGO    2083
226  DODGE               GR       3
227  DODGE            GRAND      54
228  DODGE    GRAND CARAVAN    7977
229  DODGE         INTREPID     119
230  DODGE          JOURNEY    3110
231  DODGE           MAGNUM     399
232  DODGE             NEON     247
233  DODGE            NITRO     610
234  DODGE              RAM      11
235  DODGE        RAM CARGO       8
236  DODGE  RAM PICKUP 1500    2235
237  DODGE  RAM PICKUP 2500     505
238  DODGE  RAM PICKUP 3500     200
239  DODGE          RAM3500       3
240  DODGE           SPIRIT       1
241  DODGE         SPRINTER       5
242  DODGE   SPRINTER CARGO     119
243  DODGE          STRATUS     319
244  DODGE            VIPER      11

In [712]:
df[(df['make']=='DODGE') & (df['model'].isin(['GRAND','GR']))][['model','codigo']]

model    codigo
5212       GR  1b4gp44r
10354   GRAND  2c4rdgdg
11861   GRAND  2d4rn4dx
11871   GRAND  2d4rn4dx
15385   GRAND  2d8hn44h
...       ...       ...
119155  GRAND  2b4gp44g
131209  GRAND  1b4hs28z
131212  GRAND  1b4gp44g
131395  GRAND  1b4gp44r
131437  GRAND  1b4gp54l

[93 rows x 2 columns]

In [714]:
df_1[(df_1['make']=='DODGE') & (df_1['model'].isin(['GRAND','GR']))][['model','codigo']]

model    codigo
18411     GR  1b4gp44r
18412     GR  1b4gp44r
18413     GR  1b4gp44r
24102  GRAND  2d4rn4dx
24103  GRAND  2d4rn4dx
26353  GRAND  2d8hn44h
26354  GRAND  2d8hn44h
26355  GRAND  2d8hn44h
26356  GRAND  2d8hn44h
26357  GRAND  2d8hn44h
26358  GRAND  2d8hn44h
26359  GRAND  2d8hn44h
26360  GRAND  2d8hn44h
26361  GRAND  2d8hn44h
26362  GRAND  2d8hn44h
26363  GRAND  2d8hn44h
26364  GRAND  2d8hn44h
26365  GRAND  2d8hn44h
26366  GRAND  2d8hn44h
26367  GRAND  2d8hn44h
26368  GRAND  2d8hn44h
26369  GRAND  2d8hn44h
26370  GRAND  2d8hn44h
26371  GRAND  2d8hn44h
26939  GRAND  2d8hn44p
27535  GRAND  1d4gp24e
27536  GRAND  1d4gp24e
27537  GRAND  1d4gp24e
27538  GRAND  1d4gp24e
27539  GRAND  1d4gp24e
28561  GRAND  2d4gp243
28562  GRAND  2d4gp243
28587  GRAND  1d4gp243
28588  GRAND  1d4gp243
28589  GRAND  1d4gp243
28590  GRAND  1d4gp243
28591  GRAND  1d4gp243
28592  GRAND  1d4gp243
28593  GRAND  1d4gp243
28595  GRAND  2d4gp443
28596  GRAND  2d4gp443
28597  GRAND  2d4gp443
28598  GRAND  2d4gp443
28651  GRAND  1d8gp243
28842  GRAND  2b4gp44g
28843  GRAND  2b4gp44g
28844  GRAND  2b4gp44g
28845  GRAND  2b4gp44g
28846  GRAND  2b4gp44g
28847  GRAND  2b4gp44g
28848  GRAND  2b4gp44g
29892  GRAND  1b4gp74l
29897  GRAND  1b4gp54l
29898  GRAND  1b4gp54l
29899  GRAND  1b4gp54l
29981  GRAND  1d8gp24e
30528  GRAND  1b4gp44g

In [683]:
col3_make[col3_make['make']==lista_make[13]]

make       model  seller
245  FERRARI         360       1
246  FERRARI  458 ITALIA       1
247  FERRARI  CALIFORNIA      13
248  FERRARI        F430       4

In [684]:
col3_make[col3_make['make']==lista_make[14]]

make model  seller
249  FIAT   500     557
250  FIAT  500E       3
251  FIAT  500L     305

In [685]:
col3_make[col3_make['make']==lista_make[15]]

make  model  seller
252  FISKER  KARMA       9

In [686]:
col3_make[col3_make['make']==lista_make[16]]['model'].values

array(['350', 'ASPIRE', 'BRONCO', 'C-MAX ENERGI', 'C-MAX HYBRID',
       'CONTOUR', 'CROWN', 'CROWN VICTORIA', 'E-150', 'E-SERIES VAN',
       'E-SERIES WAGON', 'E150', 'E250', 'E350', 'ECONOLINE CARGO',
       'ECONOLINE WAGON', 'EDGE', 'ESCAPE', 'ESCAPE HYBRID', 'ESCORT',
       'EXCURS', 'EXCURSION', 'EXPEDIT', 'EXPEDITION', 'EXPEDITION EL',
       'EXPLORER', 'EXPLORER SPORT', 'EXPLORER SPORT TRAC', 'F-150',
       'F-150 HERITAGE', 'F-250', 'F-250 SUPER DUTY', 'F-350 SUPER DUTY',
       'F-450 SUPER DUTY', 'F150', 'F250', 'F350', 'FIESTA',
       'FIVE HUNDRED', 'FLEX', 'FOCUS', 'FOCUS ST', 'FREESTAR',
       'FREESTYLE', 'FUSION', 'FUSION ENERGI', 'FUSION HYBRID', 'MUSTANG',
       'MUSTANG SVT COBRA', 'POLICE', 'RANGER', 'SHELBY GT500', 'TAURUS',
       'TAURUS X', 'TEMPO', 'THUNDERBIRD', 'TRANSIT CONNECT',
       'TRANSIT VAN', 'TRANSIT WAGON', 'WINDSTAR', 'WINDSTAR CARGO'],
      dtype=object)

In [687]:
col3_make[col3_make['make']==lista_make[17]]

make  model  seller
314  GEO  METRO       1
315  GEO  PRIZM      18

In [688]:
col3_make[col3_make['make']==lista_make[18]]

make                  model  seller
316  GMC                   1500       5
317  GMC                 ACADIA    2034
318  GMC                 CANYON     118
319  GMC                  ENVOY     659
320  GMC               ENVOY XL     239
321  GMC              ENVOY XUV      56
322  GMC                  JIMMY      16
323  GMC                 SAFARI       5
324  GMC           SAFARI CARGO       2
325  GMC                 SAVANA      13
326  GMC           SAVANA CARGO     662
327  GMC            SIERRA 1500    2364
328  GMC    SIERRA 1500 CLASSIC      31
329  GMC     SIERRA 1500 HYBRID       2
330  GMC          SIERRA 1500HD       6
331  GMC            SIERRA 2500       4
332  GMC          SIERRA 2500HD     440
333  GMC  SIERRA 2500HD CLASSIC      35
334  GMC            SIERRA 3500      10
335  GMC          SIERRA 3500HD      48
336  GMC                 SONOMA      57
337  GMC                     SR       2
338  GMC                 SUBRBN       1
339  GMC               SUBURBAN      20
340  GMC                TERRAIN    1136
341  GMC                  YUKON    1380
342  GMC           YUKON DENALI       4
343  GMC           YUKON HYBRID      34
344  GMC               YUKON XL    1264

In [689]:
col3_make[col3_make['make']==lista_make[19]]

make          model  seller
345  HONDA         ACCORD    9151
346  HONDA  ACCORD HYBRID      10
347  HONDA          CIVIC    7439
348  HONDA  CIVIC DEL SOL       1
349  HONDA           CR-V    3265
350  HONDA           CR-Z     108
351  HONDA      CROSSTOUR     406
352  HONDA        ELEMENT     308
353  HONDA            FIT     441
354  HONDA        INSIGHT     228
355  HONDA        ODYSSEY    3536
356  HONDA       PASSPORT      29
357  HONDA          PILOT    2149
358  HONDA        PRELUDE      14
359  HONDA      RIDGELINE     214
360  HONDA          S2000      46

In [690]:
col3_make[col3_make['make']==lista_make[21]]

make            model  seller
366  HYUNDAI           ACCENT    2040
367  HYUNDAI            AZERA     314
368  HYUNDAI          ELANTRA    6265
369  HYUNDAI    ELANTRA COUPE     103
370  HYUNDAI       ELANTRA GT     245
371  HYUNDAI  ELANTRA TOURING     127
372  HYUNDAI        ENTOURAGE      70
373  HYUNDAI            EQUUS     132
374  HYUNDAI          GENESIS     902
375  HYUNDAI    GENESIS COUPE     294
376  HYUNDAI            SANTA      16
377  HYUNDAI         SANTA FE    1602
378  HYUNDAI   SANTA FE SPORT     348
379  HYUNDAI           SONATA    6913
380  HYUNDAI    SONATA HYBRID     546
381  HYUNDAI          TIBURON     191
382  HYUNDAI           TUCSON     874
383  HYUNDAI         VELOSTER     607
384  HYUNDAI         VERACRUZ     156
385  HYUNDAI            XG300       9
386  HYUNDAI            XG350      79

In [691]:
col3_make[col3_make['make']==lista_make[22]]

make          model  seller
387  INFINITI             EX     160
388  INFINITI           EX35      47
389  INFINITI             FX     450
390  INFINITI           FX35     296
391  INFINITI           FX45      35
392  INFINITI           FX50       4
393  INFINITI  G CONVERTIBLE     385
394  INFINITI        G COUPE    1641
395  INFINITI        G SEDAN    7416
396  INFINITI            G20      39
397  INFINITI            G35    1121
398  INFINITI            G37     215
399  INFINITI            I30      73
400  INFINITI            I35      96
401  INFINITI            J30       5
402  INFINITI             JX     490
403  INFINITI              M    1025
404  INFINITI            M35     257
405  INFINITI            M45      67
406  INFINITI            M56       5
407  INFINITI            Q45      44
408  INFINITI            Q50     539
409  INFINITI             QX     216
410  INFINITI            QX4      74
411  INFINITI           QX50      10
412  INFINITI           QX56     531
413  INFINITI           QX60      15
414  INFINITI           QX70      48

In [692]:
col3_make[col3_make['make']==lista_make[23]]

make     model  seller
415  ISUZU     AMIGO       5
416  ISUZU  ASCENDER      38
417  ISUZU     AXIOM      19
418  ISUZU  I-SERIES       9
419  ISUZU     RODEO     101
420  ISUZU   TROOPER      32

In [693]:
col3_make[col3_make['make']==lista_make[24]]

make      model  seller
421  JAGUAR     F-TYPE      39
422  JAGUAR     S-TYPE     240
423  JAGUAR     X-TYPE     190
424  JAGUAR         XF     391
425  JAGUAR         XJ     213
426  JAGUAR  XJ-SERIES     175
427  JAGUAR         XK      83
428  JAGUAR  XK-SERIES      89

In [694]:
col3_make[col3_make['make']==lista_make[25]]

make           model  seller
429  JEEP        CHEROKEE     607
430  JEEP       COMMANDER     671
431  JEEP         COMPASS    1382
432  JEEP              GR       9
433  JEEP  GRAND CHEROKEE    4699
434  JEEP         LIBERTY    2556
435  JEEP         PATRIOT    1787
436  JEEP        WRANGLER    3772

In [695]:
col3_make[col3_make['make']==lista_make[26]]

make     model  seller
437  KIA    AMANTI     101
438  KIA   BORREGO      83
439  KIA   CADENZA     245
440  KIA     FORTE    2751
441  KIA      K900      61
442  KIA    OPTIMA    4389
443  KIA       RIO    1485
444  KIA     RONDO     163
445  KIA    SEDONA     946
446  KIA   SORENTO    3408
447  KIA      SOUL    2743
448  KIA   SPECTRA     484
449  KIA  SPORTAGE    1225

In [696]:
col3_make[col3_make['make']==lista_make[27]]

make     model  seller
450  LAMBORGHINI  GALLARDO       4

In [697]:
col3_make[col3_make['make']==lista_make[28]]

make                model  seller
451  LAND ROVER            DISCOVERY      28
452  LAND ROVER  DISCOVERY SERIES II      31
453  LAND ROVER           FREELANDER      50
454  LAND ROVER                  LR2     137
455  LAND ROVER                  LR3      86
456  LAND ROVER                  LR4     175
457  LAND ROVER                RANGE       5
458  LAND ROVER          RANGE ROVER     427
459  LAND ROVER   RANGE ROVER EVOQUE     151
460  LAND ROVER    RANGE ROVER SPORT     800
461  LAND ROVER                  RRS       1

In [698]:
col3_make[col3_make['make']==lista_make[29]]

make      model  seller
462  LEXUS    CT 200H     412
463  LEXUS     ES 300     381
464  LEXUS    ES 300H      72
465  LEXUS     ES 330     309
466  LEXUS     ES 350    1921
467  LEXUS     GS 300     219
468  LEXUS     GS 350     766
469  LEXUS     GS 400      20
470  LEXUS     GS 430      48
471  LEXUS    GS 450H      33
472  LEXUS     GS 460       6
473  LEXUS     GX 460     257
474  LEXUS     GX 470     203
475  LEXUS    HS 250H      72
476  LEXUS     IS 250    1939
477  LEXUS   IS 250 C     125
478  LEXUS     IS 300     112
479  LEXUS     IS 350     195
480  LEXUS   IS 350 C      17
481  LEXUS       IS F      31
482  LEXUS     LS 400      86
483  LEXUS     LS 430     386
484  LEXUS     LS 460     574
485  LEXUS  LS 600H L       4
486  LEXUS     LX 450       5
487  LEXUS     LX 470      79
488  LEXUS     LX 570     121
489  LEXUS     RC 350       6
490  LEXUS       RC F       5
491  LEXUS     RX 300     334
492  LEXUS     RX 330     411
493  LEXUS     RX 350    2379
494  LEXUS    RX 400H     164
495  LEXUS    RX 450H     178
496  LEXUS     SC 300      11
497  LEXUS     SC 400      15
498  LEXUS     SC 430      83

In [699]:
col3_make[col3_make['make']==lista_make[30]]

make        model  seller
499  LINCOLN      AVIATOR     150
500  LINCOLN    BLACKWOOD       7
501  LINCOLN  CONTINENTAL      67
502  LINCOLN           LS     218
503  LINCOLN      MARK LT      76
504  LINCOLN    MARK VIII       9
505  LINCOLN          MKC      20
506  LINCOLN          MKS     877
507  LINCOLN          MKT     172
508  LINCOLN          MKX    1160
509  LINCOLN          MKZ    1715
510  LINCOLN    NAVIGATOR     833
511  LINCOLN  NAVIGATOR L      23
512  LINCOLN         TOWN      12
513  LINCOLN     TOWN CAR     351
514  LINCOLN       ZEPHYR      95

In [700]:
col3_make[col3_make['make']==lista_make[31]]

make  model  seller
515  LOTUS  EXIGE       1

In [701]:
col3_make[col3_make['make']==lista_make[32]]

make                    model  seller
516  MASERATI                    COUPE       2
517  MASERATI                   GHIBLI      19
518  MASERATI                GRANSPORT       2
519  MASERATI              GRANTURISMO      28
520  MASERATI  GRANTURISMO CONVERTIBLE      13
521  MASERATI             QUATTROPORTE      69
522  MASERATI                   SPYDER       3

In [702]:
col3_make[col3_make['make']==lista_make[33]]

make                  model  seller
523  MAZDA                      3       2
524  MAZDA                      6      14
525  MAZDA                    626     102
526  MAZDA        B-SERIES PICKUP      10
527  MAZDA         B-SERIES TRUCK       5
528  MAZDA                   CX-5     507
529  MAZDA                   CX-7     433
530  MAZDA                   CX-9     845
531  MAZDA                 MAZDA2     418
532  MAZDA                 MAZDA3    2906
533  MAZDA                 MAZDA5     643
534  MAZDA                 MAZDA6    1531
535  MAZDA           MAZDASPEED 3      25
536  MAZDA      MAZDASPEED MAZDA3      22
537  MAZDA      MAZDASPEED MAZDA6      14
538  MAZDA  MAZDASPEED MX-5 MIATA       3
539  MAZDA     MAZDASPEED PROTEGE       5
540  MAZDA            MAZDASPEED3      33
541  MAZDA               MILLENIA      53
542  MAZDA                    MPV     155
543  MAZDA             MX-5 MIATA     227
544  MAZDA                PROTEGE     116
545  MAZDA               PROTEGE5      47
546  MAZDA                   RX-8      15
547  MAZDA                    RX8      43
548  MAZDA                TRIBUTE     322
549  MAZDA         TRIBUTE HYBRID       2
550  MAZDA                  TRUCK      10

In [703]:
col3_make[col3_make['make']==lista_make[34]]

make                   model  seller
551  MERCEDES               190-CLASS       3
552  MERCEDES               300-CLASS      12
553  MERCEDES               420-CLASS       1
554  MERCEDES                  420SEL       1
555  MERCEDES               500-CLASS       7
556  MERCEDES  B-CLASS ELECTRIC DRIVE       5
557  MERCEDES                    B200       2
558  MERCEDES                 C-CLASS    5449
559  MERCEDES                CL-CLASS     177
560  MERCEDES               CLA-CLASS     134
561  MERCEDES               CLK-CLASS     477
562  MERCEDES               CLS-CLASS     475
563  MERCEDES                 E-CLASS    3884
564  MERCEDES                  E300DT       2
565  MERCEDES                 G-CLASS      68
566  MERCEDES                GL-CLASS    1061
567  MERCEDES               GLA-CLASS       2
568  MERCEDES               GLK-CLASS     857
569  MERCEDES                 M-CLASS    1908
570  MERCEDES                 R-CLASS     332
571  MERCEDES                 S-CLASS    1518
572  MERCEDES                SL-CLASS     436
573  MERCEDES               SLK-CLASS     321
574  MERCEDES                 SLS AMG       3
575  MERCEDES              SLS AMG GT       1
576  MERCEDES                SPRINTER      74

In [704]:
col3_make[col3_make['make']==lista_make[35]]

make           model  seller
577  MERCURY          COUGAR      39
578  MERCURY   GRAND MARQUIS     552
579  MERCURY         MARINER     312
580  MERCURY  MARINER HYBRID      23
581  MERCURY           MILAN     274
582  MERCURY    MILAN HYBRID       6
583  MERCURY         MONTEGO      93
584  MERCURY        MONTEREY      48
585  MERCURY     MOUNTAINEER     424
586  MERCURY        MYSTIQUE       2
587  MERCURY           SABLE     221
588  MERCURY          TRACER       4
589  MERCURY        VILLAGER      24

In [705]:
df.columns

Index(['year', 'make', 'model', 'trim', 'body', 'transmission', 'vin', 'state',
       'condition', 'odometer', 'color', 'interior', 'seller', 'mmr',
       'sellingprice', 'saledate', 'codigo_marca', 'codigo_modelo',
       'codigo_transmision', 'codigo_trim', 'codigo'],
      dtype='object')

In [706]:
df[df['make']=='MAZDA']

year   make              model             trim       body  \
369     2013  MAZDA             MAZDA2            SPORT  Hatchback   
375     2013  MAZDA             MAZDA2          TOURING  Hatchback   
388     2013  MAZDA             MAZDA3             I SV      Sedan   
392     2013  MAZDA               CX-9          TOURING        SUV   
396     2013  MAZDA               CX-9          TOURING        SUV   
...      ...    ...                ...              ...        ...   
558645  2003  MAZDA             MAZDA6                I      sedan   
558691  2010  MAZDA             MAZDA6  I GRAND TOURING      sedan   
558705  2008  MAZDA               CX-7            SPORT        suv   
558746  2014  MAZDA             MAZDA3             I SV      sedan   
558756  2008  MAZDA  MAZDASPEED MAZDA3    GRAND TOURING  hatchback   

       transmission                vin state  condition  odometer  ...  \
369       automatic  jm1de1kz8d0152764    ca        2.0   24252.0  ...   
375       automatic  jm1de1lz0d0160159    ca       35.0   44392.0  ...   
388       automatic  jm1bl1tg8d1728177    ca       36.0   54715.0  ...   
392       automatic  jm3tb3cv9d0410440    ca       38.0   46733.0  ...   
396       automatic  jm3tb2cv8d0412139    ca       36.0   44289.0  ...   
...             ...                ...   ...        ...       ...  ...   
558645          NaN  1yvfp80c735m25070    pa       35.0   73432.0  ...   
558691    automatic  1yvhz8ch8a5m44685    fl       36.0   58814.0  ...   
558705    automatic  jm3er293080209309    fl       19.0   96672.0  ...   
558746    automatic  jm1bm1t73e1191748    nv       28.0   27190.0  ...   
558756       manual  jm1bk34l081830430    co       19.0   69731.0  ...   

       interior                        seller      mmr  sellingprice  \
369       black           exeter finance corp   9825.0        9100.0   
375       black         the hertz corporation   8475.0        8500.0   
388       black      california republic bank   9900.0       10600.0   
392       black          avis rac/san leandro  19150.0       20500.0   
396       black          avis rac/san leandro  19050.0       20000.0   
...         ...                           ...      ...           ...   
558645     gray  r hollenshead auto sales inc   2400.0        3100.0   
558691    black             ferco motors corp   8225.0        9400.0   
558705     gray          meridian remarketing   3950.0        3400.0   
558746    black            kapitolia auto llc  11650.0       11000.0   
558756    black   prestige financial services   9400.0       10200.0   

                                       saledate codigo_marca codigo_modelo  \
369     Thu Dec 18 2014 12:00:00 GMT-0800 (PST)           jm            1d   
375     Tue Dec 16 2014 12:30:00 GMT-0800 (PST)           jm            1d   
388     Wed Dec 17 2014 12:30:00 GMT-0800 (PST)           jm            1b   
392     Wed Dec 31 2014 12:30:00 GMT-0800 (PST)           jm            3t   
396     Wed Dec 31 2014 12:30:00 GMT-0800 (PST)           jm            3t   
...                                         ...          ...           ...   
558645  Fri Jun 19 2015 02:00:00 GMT-0700 (PDT)           1y            vf   
558691  Fri Jun 19 2015 02:45:00 GMT-0700 (PDT)           1y            vh   
558705  Fri Jun 19 2015 03:20:00 GMT-0700 (PDT)           jm            3e   
558746  Fri Jun 19 2015 05:00:00 GMT-0700 (PDT)           jm            1b   
558756  Wed Jun 24 2015 04:45:00 GMT-0700 (PDT)           jm            1b   

       codigo_transmision codigo_trim    codigo  
369                     e          1k  jm1de1kz  
375                     e          1l  jm1de1lz  
388                     l          1t  jm1bl1tg  
392                     b          3c  jm3tb3cv  
396                     b          2c  jm3tb2cv  
...                   ...         ...       ...  
558645                  p          80  1yvfp80c  
558691                  z          8c  1yvhz8ch  
558705                  r 